code/datums/components/usb_port.dm

```diff
@@ -0,0 +1,163 @@
+/// Opens up a USB port that can be connected to by circuits, creating registerable circuit components
+/datum/component/usb_port
+	/// The component types to create when something plugs in
+	var/list/circuit_component_types
+
+	/// The currently connected circuit
+	var/obj/item/integrated_circuit/attached_circuit
+
+	/// The currently connected USB cable
+	var/datum/weakref/usb_cable_ref
+
+	/// The components inside the parent
+	var/list/obj/item/circuit_component/circuit_components
+
+	/// The beam connecting the USB cable to the machine
+	var/datum/beam/usb_cable_beam
+
+/datum/component/usb_port/Initialize(list/circuit_component_types)
+	if (!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.circuit_component_types = circuit_component_types
+
+/datum/component/usb_port/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_USB_CABLE_TRY_ATTACH, .proc/on_atom_usb_cable_try_attach)
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/on_moved)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+
+/datum/component/usb_port/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_USB_CABLE_TRY_ATTACH,
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_PARENT_EXAMINE,
+	))
+
+	unregister_circuit_signals()
+
+/datum/component/usb_port/Destroy()
+	QDEL_LAZYLIST(circuit_components)
+	QDEL_NULL(usb_cable_beam)
+
+	attached_circuit = null
+	usb_cable_ref = null
+
+	return ..()
+
+/datum/component/usb_port/proc/unregister_circuit_signals()
+	if (isnull(attached_circuit))
+		return
+
+	UnregisterSignal(attached_circuit, list(
+		COMSIG_CIRCUIT_SHELL_REMOVED,
+		COMSIG_PARENT_QDELETING,
+	))
+
+	var/shell = attached_circuit.shell
+	if (!isnull(shell))
+		UnregisterSignal(shell, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(shell, COMSIG_PARENT_EXAMINE)
+
+/datum/component/usb_port/proc/create_circuit_components(obj/item/integrated_circuit/circuitboard)
+	var/created_circuit_components = list()
+
+	for(var/circuit_component_type in circuit_component_types)
+		var/obj/item/circuit_component/circuit_component = new circuit_component_type(parent)
+		circuit_component.insertable_through_shells = TRUE
```

Comment:
might wanna change all the true/false variables on circuit_component to one variable flag to avoid variable bloat